event_sink_clickhouse/sinks/course_published.py

```diff
@@ -0,0 +1,250 @@
+
+        Args:
+            course_id: CourseKey of the course we want to serialize
+
+        Returns:
+            nodes: a list of dicts representing xblocks for the course
+            relationships: a list of dicts representing relationships between nodes
+        """
+        modulestore = CoursePublishedSink._get_modulestore()
+        detached_xblock_types = CoursePublishedSink._get_detached_xblock_types()
+
+        dump_id = str(uuid.uuid4())
+        dump_timestamp = str(timezone.now())
+
+        # Create a location to node mapping as a lookup for writing relationships later
+        location_to_node = {}
+        items = modulestore.get_items(course_id)
+
+        # Serialize the XBlocks to dicts and map them with their location as keys the
+        # whole map needs to be completed before we can define relationships
+        index = 0
+        for item in items:
+            index += 1
+            fields = self.serialize_item(item, index, detached_xblock_types, dump_id, dump_timestamp)
+            location_to_node[self.strip_branch_and_version(item.location)] = fields
+
+        # Create a list of relationships between blocks, using their locations as identifiers
+        relationships = []
+        for item in items:
+            for index, child in enumerate(item.get_children()):
+                parent_node = location_to_node.get(self.strip_branch_and_version(item.location))
+                child_node = location_to_node.get(self.strip_branch_and_version(child.location))
+
+                if parent_node is not None and child_node is not None:
+                    relationship = {
+                        'course_key': str(course_id),
+                        'parent_location': str(parent_node["location"]),
+                        'child_location': str(child_node["location"]),
+                        'order': index,
+                        'dump_id': dump_id,
+                        'time_last_dumped': dump_timestamp,
+                    }
+                    relationships.append(relationship)
+
+        nodes = list(location_to_node.values())
+        return nodes, relationships
+
+    def _send_xblocks(self, serialized_xblocks):
+        """
+        Create the insert query and CSV to send the serialized XBlocks to ClickHouse.
+        """
+        params = CLICKHOUSE_BULK_INSERT_PARAMS.copy()
+
+        # "query" is a special param for the query, it's the best way to get the FORMAT CSV in there.
+        params["query"] = f"INSERT INTO {self.ch_database}.course_blocks FORMAT CSV"
+
+        output = io.StringIO()
+        writer = csv.writer(output, quoting=csv.QUOTE_NONNUMERIC)
+
+        for node in serialized_xblocks:
+            writer.writerow(node.values())
+
+        request = requests.Request(
+            'POST',
+            self.ch_url,
+            data=output.getvalue(),
+            params=params,
+            auth=self.ch_auth
+        )
+
+        self._send_clickhouse_request(request)
```

Comment:
is there a reason why do we create the request outside `_send_clickhouse_request`?

Reply:
I was thinking it would give us more flexibility in the future, for instance if we needed PUT requests or wanted to use params to send data instead of putting it in the body, but I honestly didn't give it a ton of thought. I'd like to see what happens with the next couple of PRs here and decide on this and the testing question once there are additional use cases.